docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,380 @@
+
+(5) In the `name` section, you can assign a name to the group, which is later used for matching when there are multiple groups. If no group name is provided, the values from the group will apply to all entries in previous groups.
+
+(6) In `fallback` section, you can configure the default behaviour in case of other zones fail, or exclude specific zones.
+
+(7) At last we should apply some fallback if no dataplanes from previous configuration are available, at this moment we allow `Fail` and `Any` fallback. Default value `Fail`'
+
+(8) In `exclude` section, you can configure which zones shouldn't take part in the routing
+
+(9) In [`overprovisioningFactor`](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/load_balancing/overprovisioning#arch-overview-load-balancing-overprovisioning-factor) section, you configure how early [Envoy] should consider other priorities.
+
+
+### API examples based on use cases
+
+#### Keeping traffic in same zone (disable cross zone traffic)
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      inZone: []
+```
+
+![Use case 1](assets/031/use_case_1.png)
+
+#### Load balancing based on priorities within a zone. Example: node → AZ → region
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      inZone: ["k8s.io/node", "k8s.io/az"]
+```
+
+![Use case 2](assets/031/use_case_2_1.png)
+![Use case 2](assets/031/use_case_2_2.png)
+![Use case 2](assets/031/use_case_2_3.png)
+
+#### Prefer traffic in same zone with fallback to any other zone
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      inZone: []
+      crossZone:
+        fallback:
+          type: Any
+```
+
+![Use case 3](assets/031/use_case_3_1.png)
+![Use case 3](assets/031/use_case_3_2.png)
+
+#### Load balance equally to list of zones
+In this case, traffic from outside to the `backend` service in `zone-2` is completely excluded.
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      inZone: []
+      crossZone:
+        - groups:
+          - name: "group1"
+            zones: ["zone-1", "zone-3"]
+        fallback:
+          type: Fail
+```
+
+![Use case 4](assets/031/use_case_4_1.png)
+![Use case 4](assets/031/use_case_4_2.png)
+![Use case 4](assets/031/use_case_4_3.png)
+
+#### When leaving zone, target specific zones first
+
+```yaml
+targetRef:
+  kind: Mesh
+to:
+  targetRef:
+    kind: MeshService
+    name: backend
+  defaults:
+    localityAwareness:
+      inZone: []
+      crossZone:
+        - groups:
+          - zones: ["zone-1"]
+            name: "group1"
+        - groups:
+          - zones: ["zone-3"]
+            name: "group1"
+```
```

Comment:
Would I achieve the same thing with this YAML
```
to:
  targetRef:
    kind: MeshService
    name: backend
  defaults:
    localityAwareness:
      inZone: []
      crossZone:
        - groups:
          - zones: ["zone-1"]
            name: "group1"
          - zones: ["zone-3"]
            name: "group2"
```
?

Reply:
No that would be different, now thinking about this I feel like `groups` are not a great idea in this shape. How would you match that only services in zone `zone-5` should route to `zone-1` but services in `zone-5` should route `zone-3`. I think we are missing a matching mechanism that would say that, or we could use matching by `MeshServiceSubset` and instead of many entries in the group at the same priority level, we could do only one List for a single priority.